articles/141_static_remapping.md

```diff
@@ -0,0 +1,342 @@
+It works by first expanding the relative name and then doing FQN replacement.
+This is also part of ROS 1 remapping.
+
+*Example:*
+
+- A node is in namespace `/ns/` and uses name `bar`
+- A rule is created to remap `bar` to `foo`
+- The node's use of `bar` is expanded to `/ns/bar`
+- Both sides of the remap rule are expanded to `/ns/bar` and `/ns/foo`
+- Because the FQN match, the final name is `/ns/foo`
+
+
+### Remap via Command Line
+A user can supply node specific remapping arguments via the command line.
+Because a process can contain multiple nodes, there must be a way to uniquely identify a node in a process.
+This is a feature of ROS 1 remapping.
+
+*Example:*
+
+- Node is an executable at `/bin/my_node`
+- User wants to change `/cat` to `/dog`
+- User types `/bin/my_node /cat:=/dog`
+
+
+### Change the Default Namespace
+The default namespace is the one in which relative names get expanded to.
+This should be changeable without affecting FQN.
+ROS 1 has this feature using either the environment variable `ROS_NAMESPACE` or the argument `__ns`.
+
+*Example:*
+
+- Node is in namespace `/ns` and uses relative name `bar`
+- User changes the default namespace to `/foo`
+- The final name is `/foo/bar`
+
+## Remapping Names in ROS 1
+Remapping is a feature that also exists in ROS 1.
+In ROS 1 remapping works by passing in [arguments](http://wiki.ros.org/Remapping%20Arguments) to each node.
+Client libraries also have APIs in code to pass remapping rules when the node is initialized.
+A remap rule consists of two names: one that should be replaced with another.
+
+ROS 1 remapping works on **Fully Qualified Names** (FQN).
+Both sides of a rule are [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#a377ff8fede7b95398fd5d1c5cd49246b).
+Before a name is remapped it is also [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#ab2eebaf734abfbdccb4122f8e24f547f).
+The name is remapped to the right side only if it exactly matches the left side of a rule.
+
+
+## Remapping rule syntax
+This is a proposal for the ROS 2 remapping rule syntax.
+It attempts to be the same as ROS 1 syntax when possible, and different where it supports new features.
+
+Requirements supported by this syntax
+- Remap One Node in a Process
+- Change a Namespace
+- Change a Basename
+- Exact FQN Replacement
+- Exact Relative Name Replacement
+- Remap via Command Line
+- Change the Default Namespace
+
+Requirements not supported
+- Change a Token
+- Pre-FQN Remapping
+
+Special Operators
+- `*` wild card that matches a single token
+- `**` wild card that matches at least one slash and zero or more tokens
+- `:=` divides the two parts of a remapping rule: matching and replacement
+- `~` expands to `/namespace/nodename` and may only be at the very beginning of a match or replacement
+- `__ns` matches the default namespace if used by itself on the match side of a rule
```

Comment:
I agree that `__ns` has to be handled prior to all remappings. Remap rules happen Post-FQN expansion, but FQN expansion requires knowing what the default namespace is.

The description of `__ns` here is confusing. It is not really a remap rule. It just looks a little bit like one because it is prefixable with `nodename:` and has a right hand side `:=/some/namespace`. I can move `__ns` to its own section away from the remapping syntax.